pico/32x/sh2soc.c

```diff
@@ -137,6 +137,11 @@ static void dmac_memcpy(struct dma_chan *chan, SH2 *sh2)
 
   if (!up || chan->tcr < 4)
     return;
+  // XXX Mars Check Program fills a 64K buffer, then copies 32K longwords from
+  // DRAM to SDRAM in 4-longword mode, which is 128K. This overwrites a comm
+  // area in SDRAM, which is why the check fails.
+  // Is this a buswidth mismatch problem? As a kludge, usw 16-bit width xfers
+  if (size == 3 && (chan->sar & 0xdf000000) == 0x04000000) size = 1;
```

Comment:
Oh please don't do this, it's wrong. The real reason Mars Check Program works on real hardware is because of cache, when the memory is overwritten by DMA CPU still uses data from cache and the test passes. But I don't think it's worth emulating the cache just for this (likely bugged) MARS Check Program, if you want to pass this test so bad better detect and patch the ROM itself.

Reply:
Hmm, a caching effect... i should have thought of it myself instead of attributing this to some flaw in the hw emulation.

No, IMHO emulating caching behaviour slows down the emulation too much, and the check program isn't important either - I just wanted to see it run through to detect more hardware emulation problems. Now that it runs through but for this one thing, I suppose the hack might just go away. I agree with patching being the better solution - i would patch the DMA register load in that case.

There's another covered caching effect in pico/32x/memory.c:1769 where writes are done to a rom area and expected to be read back from the cache... should that go away as well? The rom code doesn't work correctly without covering this somehow, though.